linkerd/drain/src/lib.rs

```diff
@@ -35,28 +37,23 @@ pub struct Signal {
 /// Watch for a drain command.
 ///
 /// This wraps another future and callback to be called when drain is triggered.
+#[pin_project]
 #[derive(Clone, Debug)]
 pub struct Watch {
+    #[pin]
     drained_tx: mpsc::Sender<Never>,
-    rx: Shared<oneshot::Receiver<()>>,
-}
-
-/// The wrapped watching `Future`.
-#[derive(Debug)]
-pub struct Watching<A, F> {
-    future: A,
-    state: State<F>,
-    watch: Watch,
-}
-
-#[derive(Debug)]
-enum State<F> {
-    Watch(F),
-    Draining,
+    rx: Shared<
+        futures::future::MapErr<
+            oneshot::Receiver<()>,
+            fn(tokio::sync::oneshot::error::RecvError) -> (),
+        >,
+    >,
```

Comment:
What's the context on dropping the errors? Does this have something to do with Clone requirements? Is the error condition still meaningful to indicate the sender was dropped? If so, should we have a marker type? If not, can we drop the Result-ness of the future entirely?

Reply:
The error only ever indicates that all senders have been dropped, and we still handle it the way we did before. IIRC, you're correct; the reason it now needs to be ignored is that it lost its `Clone` impl in tokio 0.2.

I think we could just replace this with a `then` and drop the resultiness of the future entirely, good call.

Reply:
Hmm, apparently this broke the build somehow... https://github.com/linkerd/linkerd2-proxy/pull/474/checks?check_run_id=596534255

Reply:
oh, huh, the breakage is unrelated. what the heck!